openedx_tagging/core/tagging/api.py

```diff
@@ -105,5 +121,180 @@ def tag_object(
     Raised ValueError if the proposed tags are invalid for this taxonomy.
     Preserves existing (valid) tags, adds new (valid) tags, and removes omitted (or invalid) tags.
     """
-
     return taxonomy.tag_object(tags, object_id, object_type)
+
+
+def import_tags(taxonomy: Taxonomy, tags: BytesIO, format: TaxonomyDataFormat, replace=False):
+    """
+    Imports the hierarchical tags from the given blob into the Taxonomy.
+    The blob can be CSV or JSON format.
+
+    If replace, then removes any existing child Tags linked to this taxonomy before performing the import.
+    """
+
+    # Validations
+    if taxonomy.allow_free_text:
+        raise ValueError(
+            _(
+                f"Invalid taxonomy ({taxonomy.id}): You cannot import into a free-form taxonomy."
+            )
+        )
+
+    # Read file and build the tags data to be uploaded
+    try:
+        tags_data = {}
+        tags.seek(0)
+        if format == TaxonomyDataFormat.CSV:        
+            text_tags = TextIOWrapper(tags, encoding='utf-8')
+            csv_reader = csv.DictReader(text_tags)
+            header_fields = csv_reader.fieldnames
+            if csv_fields != header_fields:
+                raise ValueError(
+                    _(
+                        f"Invalid CSV header: {header_fields}. Must be: {csv_fields}."
+                    )
+                )
+            tags_data = list(csv_reader)
+        elif format == TaxonomyDataFormat.JSON:
+            tags_data = json.load(tags)
+            if 'tags' not in tags_data:
+                raise ValueError(
+                    _(
+                        f"Invalid JSON format: Missing 'tags' list."
+                    )
+                )
+            tags_data = tags_data.get('tags')
+        else:
+            raise ValueError(
+                _(
+                    f"Invalid format: {format}"
+                )
+            )
+    except ValueError as e:
+        raise e
+    finally:
+        tags.close()
+
+
+    updated_tags = []
+
+    def create_update_tag(tag):
```

Comment:
Please try to avoid inner functions this large and complex. The `create_update_tag` function has interesting logic but it's harder to test separately because it's nested in here.